### PR TITLE
Refactor daemon shutdown

### PR DESCRIPTION
**Pull request by lzpap**
# Description of change
`shutdown()` locked the daemon and blocked reporting the running background processes, hence the node didn't have info on which processes are hanging.

This change introduces `RWMutex` instead of `Mutex` as lock in daemon. A read lock is used in 
`GetRunningBackGroundWorkers()`. `shutdown()` is separated into two stages (while preserving the original logic) so that they can use `Rlock()` and `Lock()` for shutting down workers and clearing the daemon respectively.
## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Ran with GoShimmer, shutdown is properly carried out. By manually inserting delays between shutting down the workers, the node reports which background workers are still running.

## Change checklist

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by Wollac** [comment]
I've added a test to verify the expected result.
It passes even with `-race` :+1:

**Comment by lzpap** [comment]
thanks for the test, I should have done it myself :see_no_evil:

File: daemon/daemon.go [hunk 1]
```diff
@@ -222,14 +222,21 @@ func (d *OrderedDaemon) waitGroupForLastPriority() *sync.WaitGroup {
 }
 
 func (d *OrderedDaemon) shutdown() {
-	d.lock.Lock()
-	defer d.lock.Unlock()
-
 	d.stopped.Set()
```
Comment (Wollac) [inline daemon/daemon.go:225]:
```suggestion
	if !d.stopped.SetToIf(false, true) {
		return
	}
```

In the current version you cannot shutdown twice. (It would lock and then `running` will be `false`.) Now, a double shutdown could happen. Although this is strongly discouraged, it would still be good to fix.
By atomically setting and checking `stopped` this should be fixed.

Consider adding a test for this case. Something like the following should work:
```go
d := daemon.New()
err := d.BackgroundWorker("A", func(shutdownSignal <-chan struct{}) {
	<-shutdownSignal
	// sleep longer than the grace time before shutting down
	time.Sleep(2 * graceTime)
})
require.NoError(t, err)

d.Start()
time.Sleep(graceTime)

d.Shutdown()
time.Sleep(graceTime)
d.ShutdownAndWait()
```
This should hopefully fail before and pass with the suggestion :stuck_out_tongue:

Comment (Wollac) [reply daemon/daemon.go:225]: I just realize that with this version: the second ShutdownAndWait call never blocks even though we are still not completely shut down.
So an alternative would be to introduce `stoppedOnce sync.Once` instead of `stopped`. This would lead to a block also for the second call.

Comment (Wollac) [reply daemon/daemon.go:225]: Since this is an edge-case anyways. I don't really mind which option you want go for @lzpap

Comment (lzpap) [reply daemon/daemon.go:225]: Good suggestion! Indeed the test fails before with panic because the second shutdown is trying to close the already closed `shutdownSignal` channel of the workers.

Comment (lzpap) [reply daemon/daemon.go]: @Wollac okay, let me see, I didn't notice your new comments :D

Comment (lzpap) [reply daemon/daemon.go]: > I just realize that with this version: the second ShutdownAndWait call never blocks even though we are still not completely shut down.
> So an alternative would be to introduce `stoppedOnce sync.Once` instead of `stopped`. This would lead to a block also for the second call.

I'm a bit confused about the term `block` here.
The desired behavior here is to prevent calling a second shutdown while one is in progress, right? So one option is to just return from `shutdown()` if we see that the daemon is already stopped. Is this what you mean by blocking?

Currently, `stopped` is only set to `true` in `shutdown()`, so with this conditional set you suggested we essentially reach the same `do it once` behavior, no?

Comment (Wollac) [reply daemon/daemon.go]: Yes, that is the key purpose. However there is a subtle difference in the behavior:
Let's say you call `Shutdown` and then, while we are still trying to stop everything, `ShutdownAndWait` (just like in the test).
Now the question is, what is the expected behavior of `ShutdownAndWait`:
- Is it supposed to _only_ return once everything has been successfully stopped
- or should it just cause no panic?

When using `sync.Once` we get the behavior that `ShutdownAndWait` only returns when a shutdown call has been successful. I.e. when called at the same time, `ShutdownAndWait` will block until the first shutdown call is done. This means that as a user, when  `ShutdownAndWait` returns we are 100% sure that everything is shut down.
Without blocking, it could still be the case that some other process is still busy shutting down so we have no guarantee and would need to check `IsRunning` to be 100%.

To see this you could add  `assert.False(t, d.IsRunning())` after the final `d.ShutdownAndWait()` in the test and it would fail.

I guess the "blocking" solution is a wee bit cleaner, but it would also perfectly fine with me to assume that `Shutdown()` is not concurrency safe and if it is called from multiple go routines the user needs to know what they are doing

**Comment by lzpap** [comment]
What do you say @Wollac? :point_up_2::point_up_2:
